File: interview_cake/trees_and_graphs/bfs.rb
require_relative 'binary_tree'

def bfs(node)
  queue = [node]

  while queue.length != 0
    node = queue.shift
    queue.push(node.left) if node.left != nil
    queue.push(node.right) if node.right != nil
    yield node.value
  end
end

bfs(ROOT) { |value| puts value }

File: interview_cake/trees_and_graphs/bfs.js
import { root } from './binary_tree.js'

function bfs(node, func) {
  let queue = [node];

  while (queue.length > 0) {
    let currentNode = queue.shift();
    func(currentNode.value);
    if (currentNode.left != null) {
      queue.push(currentNode.left);
    }
    if (currentNode.right != null) {
      queue.push(currentNode.right);
    }
  }
}

bfs(root, value => console.log(value));
